git/odb/tree.go

```diff
@@ -91,6 +92,63 @@ func (t *Tree) Encode(to io.Writer) (n int, err error) {
 	return
 }
 
+// Merge performs a merge operation against the given set of `*TreeEntry`'s by
+// either replacing existing tree entries of the same name, or appending new
+// entries in sub-tree order.
+//
+// It returns a copy of the tree, and performs the merge in O(n*log(n)) time.
+func (t *Tree) Merge(others ...*TreeEntry) *Tree {
+	unseen := make(map[string]*TreeEntry)
+
+	// Build a cache of name+filemode to *TreeEntry.
+	for _, other := range others {
+		key := fmt.Sprintf("%s%o", other.Name, other.Filemode)
```

Comment:
Are filemodes always the same length? It may be worth adding some kind of delimiter to remove the possibility of any potential clashes. Though this is probably only a risk if someone is creating known bad `Tree` objects.